### ISSUE TITLE
Copy-DbaDbTableData - Error when attempting to copy tables with a column of datatype Geography

### ISSUE THREAD

**Issue by carlei1978** [dataplat/dbatools#9286]
### Verified issue does not already exist?

I have searched and found no existing issue

### What error did you receive?

WARNING: [14:50:56][Copy-DbaDbTableData] Something went wrong | 'Microsoft.SqlServer.Types.SqlGeography' is an invalid user defined type, reason: no UDT attribute.

If I add the switch `-EnableException` to `Copy-DbaDbTableData`  then the detailed error is
Exception:
 Line |
97864 |          throw $records[0]
      |          ~~~~~~~~~~~~~~~~~
      | 'Microsoft.SqlServer.Types.SqlGeography' is an invalid user defined type, reason: no UDT attribute.

### Steps to Reproduce

```
$secureuser = "myuser"
$securePassword = ConvertTo-SecureString "mypassword" -AsPlainText -Force
$securecred = New-Object -TypeName System.Management.Automation.PSCredential -ArgumentList $secureuser, $securePassword
$SourceDBServer = Connect-DbaInstance -SqlInstance localhost -SqlCredential $securecred -TrustServerCertificate -NonPooledConnection -EncryptConnection
$TargetDbServer = Connect-DbaInstance -SqlInstance localhost -SqlCredential $securecred -TrustServerCertificate -NonPooledConnection -EncryptConnection
$params = @{
    SqlInstance = $SourceDBServer
    Destination = $TargetDbServer
    Database = 'ExportTest'
    DestinationDatabase = 'ImportTest'
    Table = '[dbo].[GeogTest]'
    DestinationTable = '[dbo].[GeogTest]'
    KeepIdentity = $true
    KeepNulls = $true
    Truncate = $true
    BatchSize = 10000
}
Copy-DbaDbTableData @params
```

### Please confirm that you are running the most recent version of dbatools

2.1.7

### Other details or mentions

The following T-SQL can be used to create the source table and populated with some test data:

CREATE TABLE dbo.GeogTest(
    lKey INT NOT NULL,
    sGeog GEOGRAPHY NOT NULL,
    CONSTRAINT [PK_GeogTest] PRIMARY KEY CLUSTERED(
        lKey
    )
)
GO

INSERT INTO dbo.GeogTest
VALUES(1,	0xE6100000010C4772F90FE94F414031992A1895942540),
(2,	0xE6100000010C9B030473F4102940ACC5A70018655940),
(4,	0xE6100000010CD39FFD4811D139409A99999999994B40),
(5,	0xE6100000010C03EE79FEB4811040E09EE74F1B485A40),
(6,	0xE6100000010CC976BE9F1A2F114054742497FF385A40),
(7,	0xE6100000010C325A47551384264035EF384547025440),
(8,	0xE6100000010CB35E0CE544BB12C06A32E36DA5A75A40)
GO

### What PowerShell host was used when producing this error

PowerShell Core (pwsh.exe)

### PowerShell Host Version

Name                           Value
----                           -----
PSVersion                      7.4.1
PSEdition                      Core
GitCommitId                    7.4.1
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

### SQL Server Edition and Build number

Microsoft SQL Server 2022 (RTM-CU11) (KB5032679) - 16.0.4105.2 (X64)  	Nov 14 2023 18:33:19  	Copyright (C) 2022 Microsoft Corporation 	Developer Edition (64-bit) on Linux (Ubuntu 22.04.3 LTS) <X64>

### .NET Framework Version

.NET 8.0.1

**Comment by potatoqualitee** [comment]
@claude can you pls take a look and propose a solution. It's possible that I added the geography DLL from SMO since.

**Comment by potatoqualitee** [comment]
@claude are spatial types so heavy they cant be loaded with the rest of smo in the https://github.com/dataplat/dbatools.library ? ew dont load RMO bc it takes a while and so few people use it.

**Comment by andreasjordan** [comment]
@potatoqualitee do you have plans to update the library with the types?

**Comment by potatoqualitee** [comment]
Yes, it should be already. Though I thought it was already by 2025. We're quite recent with SMO. and soon i'll be doing a library update to be even recent-er.

**Comment by andreasjordan** [comment]
Tested this against the current code in a lab, using the exact repro from the original report - same table definition, same data, same parameter splat.

**It works.** dbatools from `development` with `dbatools.library` 2026.5.3 on PowerShell 7.6:

| Instance | Version | Result |
|---|---|---|
| SQL2019 | 15.0.4430.1 | `RowsCopied=7` |
| SQL2022 | 16.0.4255.1 | `RowsCopied=7` |
| SQL2025 | 17.0.4055.5 | `RowsCopied=7` |

SQL 2022 is the version from the original report. The spatial values survive the round trip rather than just the row count - reading them back with `STAsText()` on the destination gives `POINT (10.7902 34.6243)`, matching the source.

I also checked `Write-DbaDbTableData`, which uses the same `SqlBulkCopy` path and was one of the commands the earlier proposal wanted to patch. Piping rows with a geography column through it works too, and the values arrive as `Microsoft.SqlServer.Types.SqlGeography`.

**So @potatoqualitee was right - the library covers it.** The assemblies that actually load in a dbatools session are:

```
Microsoft.Data.SqlClient    6.0.0.0    ...\dbatools.library\2026.5.3\core\lib\runtimes\win\lib\net8.0\
Microsoft.SqlServer.Server  1.0.0.0    ...\dbatools.library\2026.5.3\core\lib\
Microsoft.SqlServer.Types   17.0.0.0   ...\dbatools.library\2026.5.3\core\lib\
```

**One correction worth recording**, in case this comes back. The earlier analysis in this thread said the error happens "because the spatial types assembly was not loaded", and proposed an `Add-SpatialLibrary` helper called from each bulk copy command. That is not what the error means:

- `Microsoft.SqlServer.Types` 17.x references only `netstandard` and `Microsoft.SqlServer.Server 1.0.0.0`, and stamps `SqlGeography` with `Microsoft.SqlServer.Server.SqlUserDefinedTypeAttribute` from that separate assembly
- `Microsoft.Data.SqlClient` 6.x defines no UDT attribute type of its own, so it resolves that same shared assembly

`no UDT attribute` is an **assembly identity mismatch**, not a missing assembly - it fires when the attribute stamped on the type and the attribute SqlClient looks for come from different assemblies. The classic trigger is the old .NET Framework `Microsoft.SqlServer.Types` 14.x/15.x, whose attribute comes from `System.Data`, used against `Microsoft.Data.SqlClient`. That fits the original report exactly: dbatools 2.1.7 on PowerShell 7.4 in March 2024.

This matters because an on-demand loader would have made things *worse*: force-loading a mismatched Types version reproduces the same error rather than fixing it. The right fix was the one that happened - shipping Types together with `Microsoft.SqlServer.Server` in the library. No per-command changes are needed, and the two `claude/issue-9286-*` branches from November can be abandoned.

Suggest closing this as fixed. @carlei1978 - sorry it sat this long, but if you still have that geography table around, `Update-Module dbatools, dbatools.library` should sort it.
